### PR TITLE
Return default value when prop is not find in old version tag/edge.

### LINKED ISSUE #1154: Storaged aborts when ALTER TAG to add props
image
image

### LINKED ISSUE #1185: [bug] FETCH PROP ON failed when tag schema changed
I add a new property to a tag only to find the old vertex property can't be fetched.
This is the old vertex before tag schema changed:
image
image
Then I changed the tag schema 
image
FETCH crashed
image
**QUESYION:** `FETCH` returns different results.

### PR DESCRIPTION
**Pull request by CPWstatic**
close #1185 , close #1154

For now, the storage would be crashed when we want a new property from the latest version tag/edge. The reason is that the data was stored with old tag/edge version, while the storage can't figure out what value type the property is.

### CODE REVIEW

**Comment by dangleptr** [comment]
Where did it crashed?

**Comment by CPWstatic** [comment]
> Where did it crashed?

LOG(FATAL)

**Comment by CPWstatic** [comment]
Jenkins go!

File: src/storage/Collector.h [hunk 1]
```diff
@@ -39,28 +39,65 @@ class PropsCollector : public Collector {
                 : writer_(writer) {}
 
     void collectVid(int64_t v, const PropContext& prop) override {
-        UNUSED(prop);
+        switch (prop.type_.type) {
```
Comment (dangleptr) [inline src/storage/Collector.h]: Why not reuse RowWriter to do type converting?

File: src/storage/query/QueryBaseProcessor.inl [hunk 2]
```diff
@@ -335,12 +335,23 @@ void QueryBaseProcessor<REQ, RESP>::collectProps(RowReader* reader,
...
             auto res = RowReader::getPropByName(reader, name);
             if (!ok(res)) {
-                VLOG(1) << "Skip the bad value for prop " << name;
-                continue;
+                VLOG(1) << "Bad value for prop " << name;
+                // TODO: Should return NULL
+                auto defaultVal = RowReader::getDefaultProp(prop.type_.type);
+                if (!defaultVal.ok()) {
+                    // Should never reach here.
+                    LOG(FATAL) << "Get default value failed for " << name;
```
Comment (laura-ding) [inline src/storage/query/QueryBaseProcessor.inl]: if someone adds new value type and doesn't modify the `getDefaultProp`, here will crash.

File: src/storage/test/AdHocSchemaManager.h [hunk 3]
```diff
@@ -54,6 +56,7 @@ class AdHocSchemaManager final : public nebula::meta::SchemaManager {
 
     StatusOr<std::string> toTagName(GraphSpaceID, TagID) override {
         LOG(FATAL) << "Unimplemented";
```
Comment (laura-ding) [inline src/storage/test/AdHocSchemaManager.h]: modify `FATAL` to `ERROR`

File: src/storage/test/AdHocSchemaManager.h [hunk 4]
```diff
@@ -72,6 +75,7 @@ class AdHocSchemaManager final : public nebula::meta::SchemaManager {
 
     StatusOr<std::vector<std::string>> getAllEdge(GraphSpaceID) override {
         LOG(FATAL) << "Unimplemented";
```
Comment (laura-ding) [inline src/storage/test/AdHocSchemaManager.h]: ditto

File: src/storage/query/QueryBaseProcessor.inl (hunk 2 above)
Comment (CPWstatic) [reply src/storage/query/QueryBaseProcessor.inl]: We want it crash exactly. Someone add new type, must add the corresponding test.

Comment (laura-ding) [reply src/storage/query/QueryBaseProcessor.inl]: It might  be discovered by user

File: src/storage/query/QueryBaseProcessor.inl [hunk 5]
```diff
@@ -556,6 +565,55 @@ std::vector<Bucket> QueryBaseProcessor<REQ, RESP>::genBuckets(
...
+                    vertexSchemaResp_.emplace(tc.tagId_, std::move(respTag));
+                }
+            }
+        }
+    }
+
+    if (!edgeContexts_.empty()) {
+        for (const auto& ec : edgeContexts_) {
+            nebula::cpp2::Schema respEdge;
+            RowSetWriter rsWriter;
```
Comment (dangleptr) [inline src/storage/query/QueryBaseProcessor.inl]: We don't need it anymore.

File: src/storage/query/QueryBoundProcessor.cpp [hunk 6]
```diff
@@ -113,50 +121,13 @@ kvstore::ResultCode QueryBoundProcessor::processVertex(PartitionID partId, Verte
...
-                    edgeSchema.emplace(ec.first, std::move(respEdge));
-                }
-            }
-        }
-
-        if (!edgeSchema.empty()) {
-            resp_.set_edge_schema(std::move(edgeSchema));
-        }
+    if (!edgeSchemaResp_.empty()) {
+            resp_.set_edge_schema(std::move(edgeSchemaResp_));
```
Comment (dangleptr) [inline src/storage/query/QueryBoundProcessor.cpp]: ditto

File: src/storage/query/QueryBoundProcessor.cpp [hunk 7]
```diff
@@ -113,50 +121,13 @@ kvstore::ResultCode QueryBoundProcessor::processVertex(PartitionID partId, Verte
...
-                if (it == vertexSchema.end()) {
-                    vertexSchema.emplace(tc.tagId_, respTag);
-                }
-            }
-        }
-        if (!vertexSchema.empty()) {
-            resp_.set_vertex_schema(std::move(vertexSchema));
-        }
+    if (!vertexSchemaResp_.empty()) {
+            resp_.set_vertex_schema(std::move(vertexSchemaResp_));
```
Comment (dangleptr) [inline src/storage/query/QueryBoundProcessor.cpp]: indent?

File: src/storage/query/QueryEdgePropsProcessor.cpp [hunk 8]
```diff
@@ -26,9 +26,15 @@ kvstore::ResultCode QueryEdgePropsProcessor::collectEdgesProps(
     if (ret != kvstore::ResultCode::SUCCEEDED) {
         return ret;
     }
+
+    auto schema = edgeSchema_.find(edgeKey.edge_type);
```
Comment (dangleptr) [inline src/storage/query/QueryEdgePropsProcessor.cpp]: For QueryEdgeProcessor,  it only support one edgeType one request.  So you could pass the related edge schema directly into collectEdgesProps

Comment (CPWstatic) [reply src/storage/query/QueryEdgePropsProcessor.cpp]: Got it.

File: src/storage/query/QueryEdgePropsProcessor.cpp [hunk 9]
```diff
@@ -67,17 +69,19 @@ void QueryEdgePropsProcessor::doProcess(const cpp2::EdgePropRequest& req) {
...
     }
 
-    int32_t edgeSize = std::accumulate(edgeContexts_.cbegin(), edgeContexts_.cend(), 0,
-                                       [](int ac, auto& ec) { return ac + ec.second.size(); });
-
-    int32_t returnColumnsNum = req.get_return_columns().size() + edgeSize;
+    auto schema = edgeSchema_.find(req.edge_type);
+    if (schema == edgeSchema_.end()) {
+        LOG(ERROR) << "Not found the edge type: " << req.edge_type;
+        this->onFinished();
```
Comment (dangleptr) [inline src/storage/query/QueryEdgePropsProcessor.cpp]: We should return immediately after calling ”this->onFinished“

**Pull request merged**